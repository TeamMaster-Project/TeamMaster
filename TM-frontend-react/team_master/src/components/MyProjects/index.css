.card {
  border-radius: 25px;
  text-align: center;
}

.myproject-container {
  margin: 0px 20px;
}

.project-card {
  background-color: rgb(224, 225, 231);
  margin: 15px;
  color: rgb(0, 0, 0);
}

.project-name {
  margin: 10px;
}

.register-form-container {
  background-color: #1029b8;

  min-height: 94.3vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  font-size: 1.05rem;
}

.register-form-card {
  background-color: #fff;
  border-radius: 15px;
  width: 50vw;
  opacity: 0.9;
}
.register-form-card:hover {
  opacity: 1;
}

.register-form {
  padding: 25px;
}
.register-form:hover {
  padding: 30px;
}

.form-submit-button {
  text-align: center;
}


