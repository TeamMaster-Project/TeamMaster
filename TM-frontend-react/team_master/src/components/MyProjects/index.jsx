import React, { Component } from "react";
import { Link } from "react-router-dom";
import auth from "../../services/authService";
import { getProjects, deleteProject } from "../../services/projectService";
import { toast } from "react-toastify";
import "./index.css";
import PermissionDetails from "./permissionDetails";

class MyProjects extends Component {
  state = {
    projects: [],
    currentUser: "",

    myProjectsWithModeratorAccess: [],
    myProjectsWithMemberAccess: [],
  };

  async componentDidMount() {
    const { data: projects } = await getProjects();
    const currentUser = auth.getCurrentUser();

    var filteredProjects = [];
    var myProjectsWithModeratorAccess = [];
    var myProjectsWithMemberAccess = [];

    await projects.map((project) => {
      project.moderators.map((moderator) => {
        if (moderator.email == currentUser.email) {
          filteredProjects.push(project);
          myProjectsWithModeratorAccess.push(project);
        }
      });
      project.members.map((member) => {
        if (member.email == currentUser.email) {
          filteredProjects.push(project);
          myProjectsWithMemberAccess.push(project);
        }
      });
    });

    this.setState({
      projects: filteredProjects, //Projects Replace with personalized projects
      currentUser: currentUser,
      myProjectsWithModeratorAccess: myProjectsWithModeratorAccess,
      myProjectsWithMemberAccess: myProjectsWithMemberAccess,
    });
  }

  handleDelete = async (project) => {
    //console.log(project);//Delete Projects
    const originalProjects = this.state.projects;
    const projects = originalProjects.filter((m) => m._id !== project._id);
    this.setState({ projects: projects }); //projects object array override by projects

    try {
      await deleteProject(project._id);
    } catch (ex) {
      if (ex.response && ex.response.status === 404)
        toast.error("This Project Is Already Deleted");

      this.setState({ projects: originalProjects });
    }
  };
  render() {
    console.log(this.state.projects);
    console.log(this.state.currentUser);
    console.log(
      "myProjectsWithModeratorAccess",
      this.state.myProjectsWithModeratorAccess
    );
    console.log(
      "myProjectsWithMemberAccess",
      this.state.myProjectsWithMemberAccess
    );

    if (this.state.projects.length === 0)
      return <p>There are no Projects yet</p>;

    return (
      <div className="">
        <h2>My Projects</h2>
        <h1>
          <Link
            to="/myprojects/new"
            className="btn btn-primary"
            style={{ marginBottom: 20 }}
          >
            New Project
          </Link>
        </h1>
        <div className="myproject-container">
          <div className="row">
            <div className="col-md-3 col-sm">
              <PermissionDetails />
            </div>
            <div className="col">
              <table className="table">
                <thead>
                  <tr>
                    {/* <th>Name</th> */}
                    {/* <th>Description</th> */}
                    <th>Projects that you are assigned</th>
                    <th></th>
                    <th></th>
                  </tr>
                </thead>
                <tbody>
                  {this.state.projects.map((project) => (
                    // <tr key={project._id} className="project-card card">
                    <tr key={project._id}>
                      <td className="project-name card">
                        <Link
                          to={{
                            pathname: `/myprojects/${project._id}/${project.name}`,
                            projectName: `${project.name}`,
                            description: `${project.description}`,
                          }}
                        >
                          {project.name}
                        </Link>
                      </td>
                      <td>{project.description}</td>

                      {this.state.myProjectsWithModeratorAccess.map(
                        (filteredModeratorProject) => {
                          if (filteredModeratorProject == project) {
                            return (
                              //Show edit button only for moderators
                              <td key={filteredModeratorProject._id}>
                                <button className="btn btn-light btn-sm">
                                  <Link to={`/myprojects/${project._id}`}>
                                    Edit Project
                                  </Link>
                                </button>
                                <p style={{ color: "blue" }}>
                                  <i> You are a moderator in this project</i>
                                </p>
                              </td>
                            );
                          }
                        }
                      )}

                      {this.state.myProjectsWithModeratorAccess.map(
                        (filteredModeratorProject) => {
                          if (filteredModeratorProject == project) {
                            if (
                              filteredModeratorProject.creater ===
                              this.state.currentUser.email
                            ) {
                              return (
                                //Show Delete button only for moderators
                                <td key={filteredModeratorProject._id}>
                                  <button
                                    onClick={() => this.handleDelete(project)}
                                    className="btn btn-danger btn-sm"
                                  >
                                    Delete
                                  </button>
                                  <p style={{ color: "red" }}>
                                    <i> You created this project</i>
                                  </p>
                                </td>
                              );
                            }
                          }
                        }
                      )}

                      {this.state.myProjectsWithMemberAccess.map(
                        (filteredMemberProject) => {
                          if (filteredMemberProject == project) {
                            return (
                              //Show Delete button only for moderators
                              <td key={filteredMemberProject._id}>
                                <p style={{ color: "green" }}>
                                  <i> (Member Only. No Edit Access)</i>
                                </p>
                              </td>
                            );
                          }
                        }
                      )}
                    </tr>
                  ))}
                </tbody>
              </table>
            </div>
          </div>
        </div>
      </div>
    );
  }
}

export default MyProjects;
