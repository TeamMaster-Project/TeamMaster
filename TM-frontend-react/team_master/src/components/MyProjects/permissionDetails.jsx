import React, { Component } from "react";
class PermissionDetails extends Component {
  render() {
    return (
      <React.Fragment>
        <div
          className="card text-white bg-danger mb-3"
          style={{ maxWidth: "18rem" }}
        >
          <div className="card-header">Admins</div>
          <div className="card-body">
            <p className="card-text">
              Only the Creater/Admin has permission to DELETE project
            </p>
          </div>
        </div>
        <div
          className="card text-white bg-info mb-3"
          style={{ maxWidth: "18rem" }}
        >
          <div className="card-header">Moderators</div>
          <div className="card-body">
            <p className="card-text">
              Only the moderators has permission to EDIT project details and
              MANAGE BASKETS.
            </p>
          </div>
        </div>
        <div
          className="card text-white bg-success mb-3"
          style={{ maxWidth: "18rem" }}
        >
          <div className="card-header">Members</div>
          <div className="card-body">
            <p className="card-text">
              Members who are not tagged as moderators can EDIT TASK status only
              and move it to another basket after completing.
            </p>
          </div>
        </div>
      </React.Fragment>
    );
  }
}

export default PermissionDetails;
