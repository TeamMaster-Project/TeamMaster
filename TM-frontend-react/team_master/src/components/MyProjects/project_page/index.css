.edit-baskets {
}
.vl {
  border-left: 6px solid rgb(255, 255, 255);
  height: 59vh;
}

.center-container {
  min-height: 94.3vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  font-size: 1.05rem;
}

.register-form-card {
  background-color: #fff;
  border-radius: 15px;
  width: 50vw;
  opacity: 0.9;
}
.register-form-card:hover {
  opacity: 1;
}

.register-form {
  padding: 25px;
}
.register-form:hover {
  padding: 30px;
}

.form-submit-button {
  text-align: center;
}
