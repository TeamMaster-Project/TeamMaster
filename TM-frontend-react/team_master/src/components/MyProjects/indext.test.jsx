import React from 'react'
import {render, screen, cleanup, fireEvent } from '@testing-library/react';
import renderer from 'react-test-renderer';
import MyProjects from '.';

afterEach(()=>{
    cleanup();
})

const windowOpen = jest.fn();
window.open = windowOpen;

const MyProjectsProps = {
    currentUser: { email: "newuser@gmail.com", name: "Team Master Test User" },
};


test('Matches Snapshot',()=>{
    const tree = renderer.create(<MyProjects/>).toJSON();
    expect(tree).toMatchSnapshot();
})

describe('Testing MyProjects page UI', () => {
    it('MyProject page static elements rendering correctly', async () => {
        const { getByRole, getByText } = render(<MyProjects {...MyProjectsProps}/>);

        // const title = getByText("My Projects");
        // expect(title).toBeInTheDocument();

        // const NewProjectButton = getByRole('button', { name: "New Project"});
        // expect(NewProjectButton).toBeInTheDocument();

    });  
});

// describe('Testing registration with different combinations of inputs', () => {

//     // This test case skipped because of trying to create a new user every time the test case run
//     it.skip('Sign in with correct / new credentials', async () => {
//         const { getByRole } = render(<Register {...RegisterFormPropsValid} />);
//         const RegisterButton = getByRole('button', { name: 'Register' });
//         fireEvent.click(RegisterButton);
//         await screen.findByText("Please wait...");
//         screen.findByText("Registration success");
//     });  

//     it('Trying to Sign in with correct an existing registered user', async () => {
//         const { getByRole } = render(<Register {...RegisterFormPropsWithExistingUsername} />);
//         const RegisterButton = getByRole('button', { name: 'Register' });
//         fireEvent.click(RegisterButton);
//         await screen.findByText("Please wait...");
//         screen.findByText("Registration fail");
//     });  
// })