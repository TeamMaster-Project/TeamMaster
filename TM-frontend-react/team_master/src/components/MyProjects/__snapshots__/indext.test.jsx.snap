// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches Snapshot 1`] = `
<div
  className="an_home_section_1"
>
  <div
    className="shadow-box"
  />
  <img
    height="100px"
    src="loader1.gif"
    width="100px"
  />
</div>
`;
